omniledger/service/transaction.go

```diff
@@ -67,14 +77,38 @@ type InstanceID struct {
 	SubID SubID
 }
 
-// SubID is a 32-byte id.
-type SubID [32]byte
+// NewInstanceID returns a new InstanceID given a slice of bytes.
+func NewInstanceID(buf []byte) InstanceID {
+	if len(buf) != 64 {
+		return InstanceID{ZeroDarc, ZeroSubID}
+	}
+	return InstanceID{buf[0:32], NewSubID(buf[32:64])}
+}
 
 // Slice returns concatenated DarcID and InstanceID.
-func (oid InstanceID) Slice() []byte {
+func (iID InstanceID) Slice() []byte {
 	var out []byte
-	out = append(out, oid.DarcID[:]...)
-	return append(out, oid.SubID[:]...)
+	out = append(out, iID.DarcID[:]...)
+	return append(out, iID.SubID[:]...)
+}
+
+// Equal returns if both InstanceIDs point to the same instance.
+func (iID InstanceID) Equal(other InstanceID) bool {
+	return bytes.Compare(iID.DarcID, other.DarcID) == 0 &&
+		bytes.Compare(iID.SubID[:], other.SubID[:]) == 0
+}
+
+// SubID is a 32-byte id.
+type SubID [32]byte
+
+// NewSubID returns a SubID given a slice of bytes.
+func NewSubID(buf []byte) SubID {
+	if len(buf) != 32 {
```

Comment:
It seems dangerous to silently return a subid that's zero filled when they thought they were getting a correct subid. Imagine how hard to debug if I call someone, I get a []byte back from them with len()==32, and then I send it on to NewSubID. Later the routine I called starts returning 34 bytes because of reasons. Now I am getting zeroSubID but I didn't even get a runtime warning about it.

Reply:
panic? Change the signature to add an `error` to the return?

Reply:
Or change the doc and the impl to cleary say, "Will copy len(buf) or 32 bytes, whichever is smaller." And then NewSubID(nil) does something that's valid and even documented: make a zero SubID.

Reply:
as we defined `Nonce` and `SubID` as arrays, we can in fact make it much more simple: `SubID{}` and `Nonce{}`.
Now I removed `Zero*`, except for `ZeroDarc`, because the `Darc.ID` is a slice...